### PR TITLE
[HEP] HEP Purpose, Policy, and Guidelines

### PR DESCRIPTION
**Pull request by droumis**
This PR introduces HEP 0, which outlines the purpose, policy, and guidelines for future HEPs within the HoloViz ecosystem. This document aims to clarify the process of proposing, discussing, and implementing enhancements for HoloViz evolution.

I've also updated HEP 1 to align with the format proposed by HEP 0.

### CODE REVIEW

**Review by maximlt — COMMENTED** [review]
Thanks for putting this together!

The document says a HEP `proposes a major change to the HoloViz ecosystem`. I think that's a bit of a strong statement. For example, I don't consider HEP 1 has brought any major changes, and I don't think HEP 2 I'm working on is going to change HoloViz significantly. The scope of NEPs for example is a bit more laxed with a NEP being a `design document providing information to the NumPy community, or describing a new feature for NumPy or its processes or environment`. I'd suggest widening the scope of HEPs to something similar to NEPs.

Other than that, it looks good to me.

**Comment by droumis** [comment]
> I'd suggest widening the scope of HEPs to something similar to NEPs.

I've adjusted the language to broaden the scope, please check if this is sufficient.

**Review by maximlt — APPROVED** [review]
Yes, thanks!

**Comment by maximlt** [comment]
Ah maybe another thing, perhaps the HEP should be formatted like it's saying HEPs should be, i.e. with the table and a summary?

image

**Comment by droumis** [comment]
updated to align with proposed format

**Comment by maximlt** [comment]
Thanks!

I think ultimately it'd be good to have the HEPs be visible directly on the website (like others do with NEPs, PEPs, etc.). Looking at the now formatted HEP, it has no title which won't help creating a web page out of it (I think the page title for all HEPs would be `Summary`?). Could we remove `Identifer` and `Title` from the table, and ask that the HEP title should be formatted as `# {Identifier} - {Title}`? Like [NEPs](https://numpy.org/neps/nep-0045-c_style_guide.html).

image

File: hep/hep0.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Summary
+
+HEP 0 outlines the purpose, policy, and guidelines for future HEPs. This document aims to ensure that the process of proposing and implementing enhancements to the HoloViz ecosystem is clear, structured, and inclusive.
+
+### What is a HEP?
+
+A HEP is a HoloViz Enhancement Proposal, a design document that provides information to the HoloViz community or proposes a change for which widespread attention and consensus is warranted. These changes can be technical or address the social aspects of the organization, like governance or conduct. HEPs should provide a concise specification and rationale for the proposed change. HEPs serve as the principal means for proposing significant changes to the HoloViz ecosystem, enabling the community to engage in informed decision-making while also documenting the proposals and the processes behind these decisions.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
A HEP is a HoloViz Enhancement Proposal, a design document requiring widespread attention and discussion that either proposes a specific change or else clarifies and codifies existing practice. These changes or codifications can be technical or address the social aspects of the organization, like governance or conduct. HEPs should provide a concise specification and rationale for the proposed change. HEPs serve as the principal means for proposing significant changes and formalizations for the HoloViz ecosystem, enabling the community to engage in informed decision-making while also documenting the proposals and the processes behind these decisions.
```

File: hep/hep0.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+HEPs are maintained as text files using Markdown for formatting in the [HoloViz/HoloViz](https://github.com/holoviz/holoviz) repository. Their revision history is a historical record of the proposed change.
+
+## Policy
+### Core Developers
+There are several references in this HEP to 'core developers'. This refers to the currently active HoloViz core team members, collectively described by the combined MEMBERS.md files ([example](https://github.com/holoviz/holoviz/blob/main/doc/governance/project-docs/MEMBERS.md
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
There are several references in this HEP to 'core developers'. This term refers to the currently active HoloViz core team members, collectively described by the combined MEMBERS.md files ([example](https://github.com/holoviz/holoviz/blob/main/doc/governance/project-docs/MEMBERS.md
```

File: hep/hep0.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+)) in each of the HoloViz repositories. The core developers are reachable on HoloViz [GitHub](https://github.com/holoviz/holoviz) or [Discord](https://discord.gg/rb6gPXbdAr) with the handle: `@holoviz-dev`.
+
+### Steering Committee
+There are several references in this HEP to the 'steering committee'. The currently active HoloViz steering committee is in [STEERING-COMMITTEE.md](https://github.com/holoviz/holoviz/blob/main/doc/governance/org-docs/STEERING-COMMITTEE.md), and is governed by [CHARTER.md](https://github.com/holoviz/holoviz/blob/main/doc/governance/org-docs/CHARTER.md). The steering committee members are reachable on HoloViz [GitHub](https://github.com/holoviz/holoviz) or [Discord](https://discord.gg/rb6gPXbdAr) with the handle: `@steering-committee`.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
There are several references in this HEP to the 'steering committee'. The currently active HoloViz steering committee is listed in [STEERING-COMMITTEE.md](https://github.com/holoviz/holoviz/blob/main/doc/governance/org-docs/STEERING-COMMITTEE.md), and is governed by [CHARTER.md](https://github.com/holoviz/holoviz/blob/main/doc/governance/org-docs/CHARTER.md). The steering committee members are reachable on HoloViz [GitHub](https://github.com/holoviz/holoviz) or [Discord](https://discord.gg/rb6gPXbdAr) with the handle: `@steering-committee`.
```

File: hep/hep0.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+### Workflow
+#### Initiation
+The HEP process begins with an idea for a change in how HoloViz works. These changes can be technical or address the social aspects of the organization, like governance or conduct.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
The HEP process begins with an idea for how HoloViz works or should work, whether that is a change from current practice or an explicit declaration of current practice. These enhancements can be technical or address the social aspects of the organization, such as governance or conduct. 
```

File: hep/hep0.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+### Workflow
+#### Initiation
+The HEP process begins with an idea for a change in how HoloViz works. These changes can be technical or address the social aspects of the organization, like governance or conduct. 
+
+Before beginning to write a HEP, it is best to ascertain if the idea fits the intent of the HEP process and has a chance of acceptance. Informally vetting an idea publicly before going as far as writing a HEP is meant to save the potential proposer time. Small or uncontroversial changes often do not need a HEP and can be decided on in the [community channels](https://holoviz.org/community.html). Major or potentially controversial changes, such as those that require collaboration across multiple HoloViz projects, should be submitted as HEPs. When it is unclear if a change is suited for a HEP, ask HoloViz core developers to provide guidance. For example, the 'new-contributors' Discord channel is open to the public and a good place to ask.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
Before beginning to write a HEP, it is best to ascertain if the idea fits the intent of the HEP process and has a chance of acceptance. Informally vetting an idea publicly before going as far as writing a HEP is meant to save the potential proposer time. Small or uncontroversial changes often do not need a HEP and can be decided on in the [community channels](https://holoviz.org/community.html). Major or potentially controversial changes, such as those that require collaboration across multiple HoloViz projects, are expected to be submitted as HEPs. When it is unclear if an enhancement is suited for a HEP, ask HoloViz core developers to provide guidance. For example, the 'new-contributors' Discord channel is open to the public and a good place to ask.
```

File: hep/hep0.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+#### Drafting
+Once the proposer has confirmed with the HoloViz core developers that an idea has any chance of acceptance, a draft HEP should be written following the format described below. Drafting can take place in a pull request to the HoloViz/HoloViz repository, but the status should be changed to 'draft' while working.
+
+HEPs should focus on a single issue; broad changes should be split into multiple well-focused HEPs.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
HEPs should focus on a single issue; broad topics for discussion should be split into multiple well-focused HEPs.
```

File: hep/hep0.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+#### Submission and Consensus-Building 
+Once complete, this draft should be submitted as a pull request to the HoloViz/HoloViz repository with the status changed to 'proposed'. At this point, all members of the HoloViz community are welcome and encouraged to participate in the review in a civil and respectful manner.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
Once complete, the draft HEP should be submitted as a pull request to the HoloViz/HoloViz repository with the status changed to 'proposed'. At this point, all members of the HoloViz community are welcome and encouraged to participate in the review in a civil and respectful manner.
```

File: hep/hep0.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+The HEP proposer is responsible for following the discussion on the pull request, answering questions, making updates to the proposal as part of the consensus-building process, and documenting dissenting opinions. In general, the goal is to make sure that the community has consensus, not provide a rigid policy for people to try to game. When in doubt, err on the side of asking for more feedback and looking for opportunities to compromise.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
The HEP proposer is responsible for following the discussion on the pull request, answering questions, making updates to the proposal as part of the consensus-building process, and documenting dissenting opinions. In general, the goal is to make sure that the community reaches consensus, not to provide a rigid policy for people to try to game. When in doubt, err on the side of asking for more feedback and looking for opportunities to compromise.
```

File: hep/hep0.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+When the proposer determines that the HEP is ready, they must notify the HoloViz core developers that it is ready for a formal review (e.g., in a comment on the pull request, using `@holoviz-dev`). This message must also describe any major points of contention and how they were resolved. 
+
+#### Review and Resolution
+A HoloViz core developer will then formally review the pull request for format, scope, and evidence of consensus. They may also notify other interested parties to weigh in on the HEP. If consensus is reached, the HoloViz core developer will update the status of the pull request, assign a HEP number, add a link to the discussion in the [HEP table](accepted_heps.md), and merge it. All HEPs will be resolved as either 'rejected', 'accepted', or 'deferred' depending on the consensus of the community.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
A HoloViz core developer will then formally review the pull request for format, scope, and evidence of consensus. They may also notify other interested parties to weigh in on the HEP. If consensus has been reached, the HoloViz core developer will update the status of the pull request, assign a HEP number, add a link to the discussion in the [HEP table](accepted_heps.md), and merge it. All HEPs will be resolved as either 'rejected', 'accepted', or 'deferred' depending on the consensus of the community.
```

File: hep/hep0.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+In unusual and controversial cases, a HoloViz core developer may ask the steering committee to take a HEP up for vote, following the policies described in [CHARTER.md](https://github.com/holoviz/holoviz/blob/main/doc/governance/org-docs/CHARTER.md).
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
In unusual and controversial cases where consensus has not been possible to achieve, a HoloViz core developer may ask the steering committee to take a HEP up for vote, following the policies described in [CHARTER.md](https://github.com/holoviz/holoviz/blob/main/doc/governance/org-docs/CHARTER.md).
```

File: hep/hep0.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+<tr><td> Author(s) </td><td> Full Name [email is optional] </td></tr>
+<tr><td> Created </td><td> March 15, 2024</td></tr>
+<tr><td> Updated </td><td> March 15, 2024</td></tr>
+<tr><td> Discussion </td><td> link to the issue where the HEP is being discussed, NA before submission </td></tr>
+<tr><td> Implementation </td><td> link to the PR for the implementation, NA if not available </td></tr>
+</table>
+
+This table should be followed by a **Summary** section and then
+additional sections as needed by the proposal. Some sections that may be
+included if appropriate for the proposal include.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
included if appropriate for the proposal include:
```

File: hep/hep0.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+    * Specification/Policy -- The technical details of the proposed change.
+    * Motivation -- Why the proposed change is needed.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
    * Motivation -- Why the proposed change or formalization is needed.
```

File: hep/hep0.md [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+<tr><td> Updated </td><td> March 15, 2024</td></tr>
+<tr><td> Discussion </td><td> link to the issue where the HEP is being discussed, NA before submission </td></tr>
+<tr><td> Implementation </td><td> link to the PR for the implementation, NA if not available </td></tr>
+</table>
+
+This table should be followed by a **Summary** section and then
+additional sections as needed by the proposal. Some sections that may be
+included if appropriate for the proposal include.
+
+    * Specification/Policy -- The technical details of the proposed change.
```
Comment (jbednar) [inline hep/hep0.md]:
```suggestion
    * Specification/Policy -- The technical details of the proposed enhancement.
```

File: hep/hep1.md [hunk 14]
```diff
@@ -94,14 +110,7 @@ release (3.2.0) will support minimum Python 3.9. All HoloViz projects (except
 Colorcet and Param which are covered above) should
 drop support for Python 3.7 as soon as possible to support a minimum Python of
 3.8. Following release of Bokeh 3.2.0 there should be one further release per
-HoloViz project that supports Python 3.8 and then the minimum Python version
-for each should increase to 3.9.
-
-## Revisions
-
-Changes to this document shall be recorded below:
+HoloViz project that supports Python
```
Comment (jbednar) [inline hep/hep1.md]:
```suggestion
HoloViz project that supports Python.
```

**Review by jbednar — APPROVED** [review]
Looks good! I've made a series of changes you can consider that treats "enhancement" as being _either_ a change _or_ a formalization/codification of existing practice to improve clarity or streamline processes.  Accordingly, numerous instances of the word "change" were replaced with "enhancement" or one of the two specific alternatives. 

If people find this approach awkward, I'm perfectly ok with a section at the front that says that in this document a "change" could either be an actual change in policy or practice, or simply a change in how well something has been documented or formalized. But if you do that, using "change" as the universal term for an HEP, then I'd argue that it should be "HCP", i.e., that all of them are change requests. So I'm honoring the "E" in "HEP" by  treating all of them as "enhancements", only some of which are actual changes. Someone else can decide if they agree with that approach!

**Comment by droumis** [comment]
@maximlt 

> Looking at the now formatted HEP, it has no title which won't help creating a web page out of it (I think the page title for all HEPs would be Summary?). Could we remove Identifer and Title from the table, and ask that the HEP title should be formatted as # {Identifier} - {Title}?

I updated the HEPs and the template to include a title. I would prefer to leave those rows in the table.

**Comment by maximlt** [comment]
> @maximlt
> 
> > Looking at the now formatted HEP, it has no title which won't help creating a web page out of it (I think the page title for all HEPs would be Summary?). Could we remove Identifer and Title from the table, and ask that the HEP title should be formatted as # {Identifier} - {Title}?
> 
> I updated the HEPs and the template to include a title. I would prefer to leave those rows in the table.

Ok, fine by me!

**Comment by ahuang11** [comment]
Looks mostly good; couple comments:

Is HEP1 missing: `* Resolution -- A short summary of the decision made by the community.`

Can we also put `## Resolution / Decision` as a section, or at least within Summary, e.g. precisely what was adopted?

Lastly, how do we ensure that HoloViz maintainers have seen this? When I joined, I wasn't aware of HEPs.

**Comment by droumis** [comment]
> Is HEP1 missing: * Resolution -- A short summary of the decision made by the community.
> Can we also put ## Resolution / Decision as a section, or at least within Summary, e.g. precisely what was adopted?

Currently, the resolution section is only recommended, not required. For HEP 1, I personally think the link to the GitHub PR in the table provides sufficient context for the community discussion that took place, which wasn't too much. But I've gone ahead and added a Resolution section under the summary for HEP 1. 

> Lastly, how do we ensure that HoloViz maintainers have seen this? When I joined, I wasn't aware of HEPs.

Maybe we can link to this HEP 0 from the governance documents. All HoloViz maintainers should read the governance to which they are bound, so it seems like a good place to make any new-comers aware of HEPs. But I think this is beyond discussion of HEP 0.

**Comment by ahuang11** [comment]
> Currently, the resolution section is only recommended, not required

I'm leaning towards it should be required, or at least a `Specification` on what maintainers should be doing in the future.

I imagine in the future there could be 100 HEPs; someone will not be able to read all that and their discussions. so, having a section that clearly outlines what each HEP is putting forward or whats required of maintainers would make things a whole lot more manageable.

> Maybe we can link to this HEP 0 from the governance documents

:) No one really told me anything about the governance docs either...

**Comment by maximlt** [comment]
> Lastly, how do we ensure that HoloViz maintainers have seen this? When I joined, I wasn't aware of HEPs.

I wrote https://holoviz.org/contributing.html#operating-guide for HoloViz maintainers, I think it could be updated to introduce HEPs and point to HEPs relevant to maintainers.

**Comment by droumis** [comment]
> I'm leaning towards it should be required, or at least a `Specification` on what maintainers should be doing in the future.

> I imagine in the future there could be 100 HEPs; someone will not be able to read all that and their discussions. so, having a section that clearly outlines what each HEP is putting forward or whats required of maintainers would make things a whole lot more manageable.

OK, I'll make the `Specification/Policy` required for prescriptive HEPs. But in order to understand what "precisely what was adopted", I don't think we can impose making it into a short summary. Besides, I don't think it's intended that a new contributor read through 100 HEPs, but rather that the HEPs are accessible enough and their topic sufficiently-encapsulated by the title, such that if a question arises at a HoloViz meeting (e.g. should we drop support for python X.X), that someone will remember that there is a HEP on this topic that can be used as guidance.

**Comment by maximlt** [comment]
I will be updating https://github.com/holoviz/holoviz/pull/388 in light of the proposed changes here. I think it'll help us gauge whether authoring a HEP hard following the format isn't too hard (I hope not! It shouldn't be, otherwise people will take other approaches).

**Comment by droumis** [comment]
@maximlt , since this hasn't been merged yet, please provide feedback as you update #388 so we can adjust accordingly

File: hep/hep0.md [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+- **Motivation/Background**: Why the proposed change or formalization is needed.
+- **Rationale**: Why particular decisions were made in the proposal.
+- **Backwards Compatibility**: Will the proposed change break existing packages or workflows.
+- **Alternatives**: Any alternatives considered during the design.
+- **Sample Implementation/Illustration**: Links to prototype or a sample implementation of the proposed change.
+- **FAQ**: Frequently asked questions (and answers to them).
+- **Resolution**: A short summary of the decision process conducted with the community.
+- **Reference**: Any references used in the design of the HEP.
+
```
Comment (jlstevens) [inline hep/hep0.md:96]: Maybe **Enforcement** (especially encouraging automation of enforcement e.g. in CI)? 

I suspect a lot of HEPs are proposals that we agree on but may forget to enforce appropriately if no mechanism is in place to ensure compliance. Maybe the default policy is 'memorize and review all HEPs before release' but that seems to be a lot to ask of maintainers imho...

**Comment by maximlt** [comment]
> @maximlt , since this hasn't been merged yet, please provide feedback as you update #388 so we can adjust accordingly

I updated it, mostly moving around text, and stuffing most of it under `Specification`. If you have time to have a quick look at HEP2 and see if I did that in a correct-ish way?

File: hep/hep0.md (hunk 15 above)
Comment (droumis) [reply hep/hep0.md:96]: good idea. I added a 'compliance' suggested section.

**Comment by droumis** [comment]
thanks for the input, team. merging now

**Pull request merged**